### PR TITLE
ruby things to handle the needs of bento pipeline

### PR DESCRIPTION
**Pull request by knightsamar**
All things related to building the bento pipeline will go in here.

This is a WIP PR and I will link to the complete flow diagram here soon too.

### CODE REVIEW

**Comment by DesertBeagle** [comment]
For the deeper algorithm we could just check and see if the filename for the script is inside the json file?

**Comment by DesertBeagle** [comment]
@knightsamar Shouldn't we make this a class like `Bumpzone` and `Checkzone`?

**Comment by knightsamar** [comment]
@DesertBeagle yep, that's a good idea. Please go ahead and do that. You can include the `payload_process` part in this PR.

We could also move the `octokit` part common between bumpzone, and this job to a Mixin(?) and use it so that the code is organized better.

**Comment by DesertBeagle** [comment]
@knightsamar Check out what I did for the payload processor!

File: spec/webhooks/packerpipeline_spec.rb [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+    end
+  end
+  context '#start' do
+    let(:github_mock) { double('Octokit', commits: [], issues: [], same_options?: false, auto_paginate: true) }
+    before :each do
+      allow(Octokit::Client).to receive(:new) { github_mock }
+      allow(ENV).to receive(:[])
+      allow(ENV).to receive(:[]).with('PACKER_TEMPLATES_DIR').and_return(fixture_path('').chop)
+    end
+    it 'ouputs template files put into it' do
```
Comment (knightsamar) [inline spec/webhooks/packerpipeline_spec.rb]: I think you mean to say "outputs name of template files"

File: spec/webhooks/packerpipeline_spec.rb [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+  c.include SpecHelper
+end
+
+describe PackerPipeline do
+  context '#changed_files' do
+    let(:github_mock) { double('Octokit', commits: [], issues: [], same_options?: false, auto_paginate: true) }
+    before :each do
+      allow(Octokit::Client).to receive(:new) { github_mock }
+    end
+    it 'finds single changed file' do
```
Comment (knightsamar) [inline spec/webhooks/packerpipeline_spec.rb]: The test should rather be restructured as

1. "finds changed template"
1. "finds a changed script"
1. "finds changed template and a changed script".

The `PR#`, and `arch` output test are good and should stay.
The arch output test could be expanded to process and output only one of the `arch`s at a time and then both.

File: spec/webhooks/packerpipeline_spec.rb (hunk 1 above)
Comment (DesertBeagle) [inline spec/webhooks/packerpipeline_spec.rb]: Well, yes. But I realize now that it should be outputting the image name.

Comment (knightsamar) [inline spec/webhooks/packerpipeline_spec.rb]: Actually, we should be outputting the template file names and not the image name.

**Comment by DesertBeagle** [comment]
@knightsamar Fixed output and added executable.

**Comment by DesertBeagle** [comment]
@ramereth It is ready for review!

**Comment by DesertBeagle** [comment]
[Script to receive the hook payload](https://trello.com/c/HC65bWfE/462-script-to-receive-the-hook-payload)

File: files/default/lib/packer_pipeline.rb [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+    templates = PackerPipeline.changed_files(d).map do |f|
+      find_dependent_templates(f.filename)
+    end.reduce(:+).uniq
+
+    output = {
+      'pr' => d['number']
+    }
+
+    %w(ppc64 x86_64).each do |arch|
+      output[arch] = templates.select { |t| t.include? arch }.to_ary
```
Comment (DesertBeagle) [inline files/default/lib/packer_pipeline.rb]: Why does this need a `.to_ary` on the end?

File: files/default/lib/packer_pipeline.rb [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+    end
+    output
```
Comment (DesertBeagle) [inline files/default/lib/packer_pipeline.rb]: This breaks all of the current tests. I will have to rewrite them.

Comment (DesertBeagle) [inline files/default/lib/packer_pipeline.rb]: Fixed in commit below.

**Comment by ramereth** [comment]
Could you please add more comments about each step so it's easier to understand what's going on?

**Comment by DesertBeagle** [comment]
@ramereth Added comments to PackerPipeline object!

File: spec/webhooks/packer_pipeline_spec.rb [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+      allow(ENV).to receive(:[]).with('PACKER_TEMPLATES_DIR').and_return(fixture_path(''))
+    end
+    it 'prints the PR number' do
+      file = fixture_path('osuosl.sh')
+      response_body = [double('Sawyer::Resource', filename: file)]
+      allow(github_mock).to receive(:pull_request_files).with('osuosl/packer-templates', 16).and_return(response_body)
+      allow(STDIN).to receive(:read).and_return(open_fixture('sync_packer_templates.json'))
+      expect { puts PackerPipeline.start.to_json }.to output(/16/).to_stdout
+    end
+    it 'ouputs name of template files' do
```
Comment (ramereth) [inline spec/webhooks/packer_pipeline_spec.rb]: typo

**Comment by DesertBeagle** [comment]
@ramereth Fixed!

**Comment by knightsamar** [comment]
@DesertBeagle rubocop yells! :)

**Comment by DesertBeagle** [comment]
Fixed rubocop issues!

**Comment by knightsamar** [comment]
@ramereth ready for review!

File: files/default/lib/packer_pipeline.rb [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+    Dir.glob('*.json') do |t|
+      t_data = JSON.parse(File.read(t))
+
+      # .dig fails if it tries to index an array with a string, which is ridiculous.
+      # This will fail on other json files that start with in an array.
+      # I suppose this helps more in testing than anything, since there aren't going
+      # to be other json files in the repo.
+      next unless t_data.class.name == 'Hash'
+
+      scripts = t_data.dig('provisioners', 0, 'scripts')
```
Comment (ramereth) [inline files/default/lib/packer_pipeline.rb]: I realize that 99% of our current packer template files will work with this, I know [one](https://github.com/osuosl/packer-templates/blob/master/centos-7.3-x86_64-openstack-mitaka-aio.json) that won't. Can you please make sure it works? If it were me, I would scan the array and look for ones that have ``type: shell``. Those are the only provisioners that should include ``scripts``.

File: files/default/lib/packer_pipeline.rb [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+
+      # .dig returns nil if it doesn't find that path in the hash.
+      next if scripts.nil?
+
+      dependent_templates << t if scripts.any? { |f| f.include? file }
```
Comment (ramereth) [inline files/default/lib/packer_pipeline.rb]: Comment describing what this is doing please.

File: files/default/lib/packer_pipeline.rb [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+
+  def self.start
+    d = JSON.parse(STDIN.read)
+
+    # Iterate through all the changed files and get an array of affected templates.
+    # find_dependent_templates always returns an array of strings, so I use
+    # .reduce to stitch those arrays together.
+    templates = PackerPipeline.changed_files(d).map do |f|
+      find_dependent_templates(f.filename)
+    end.reduce(:+).uniq
```
Comment (ramereth) [inline files/default/lib/packer_pipeline.rb]: What does ``:+`` do in this case?

File: spec/fixtures/openstack.sh [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+sed -i -e 's/name: fedora/name: centos/' /etc/cloud/cloud.cfg
+sed -i -e 's/Fedora Cloud User/CentOS Cloud User/' /etc/cloud/cloud.cfg
+sed -i -e 's/distro: fedora/distro: rhel/' /etc/cloud/cloud.cfg
+
+# Ensure that cloud-init starts before sshd
+if [ -e /usr/lib/systemd/system/cloud-init.service ] ; then
+  FILE=/usr/lib/systemd/system/cloud-init.service
+  sed -i '/^Wants/s/$/ sshd.service/' $FILE
+  grep -q Before $FILE && sed -i '/Before/s/$/ sshd.service/' $FILE ||  sed -i '/\[Unit\]/aBefore=sshd.service' $FILE
+fi
```
Comment (ramereth) [inline spec/fixtures/openstack.sh]: IMO you don't need a full shell script for testing fixtures. You can merely have an empty file.

File: spec/fixtures/osuosl.sh [hunk 10]
```diff
@@ -0,0 +1,10 @@
+#!/bin/bash -eux
+
+# Use OSL for repos
+sed -i -e 's/^\(mirrorlist.*\)/#\1/g' /etc/yum.repos.d/CentOS-Base.repo
+sed -i -e 's/^#baseurl=.*$releasever\/os\/$basearch\//baseurl=http\:\/\/centos.osuosl.org\/$releasever\/os\/$basearch\//g' /etc/yum.repos.d/CentOS-Base.repo
+sed -i -e 's/^#baseurl=.*$releasever\/updates\/$basearch\//baseurl=http\:\/\/centos.osuosl.org\/$releasever\/updates\/$basearch\//g' /etc/yum.repos.d/CentOS-Base.repo
+sed -i -e 's/^#baseurl=.*$releasever\/addons\/$basearch\//baseurl=http\:\/\/centos.osuosl.org\/$releasever\/addons\/$basearch\//g' /etc/yum.repos.d/CentOS-Base.repo
+sed -i -e 's/^#baseurl=.*$releasever\/extras\/$basearch\//baseurl=http\:\/\/centos.osuosl.org\/$releasever\/extras\/$basearch\//g' /etc/yum.repos.d/CentOS-Base.repo
+sed -i -e 's/^#baseurl=.*$releasever\/centosplus\/$basearch\//baseurl=http\:\/\/centos.osuosl.org\/$releasever\/centosplus\/$basearch\//g' /etc/yum.repos.d/CentOS-Base.repo
+sed -i -e 's/^#baseurl=.*$releasever\/contrib\/$basearch\//baseurl=http\:\/\/centos.osuosl.org\/$releasever\/contrib\/$basearch\//g' /etc/yum.repos.d/CentOS-Base.repo
```
Comment (ramereth) [inline spec/fixtures/osuosl.sh]: IMO you don't need a full shell script for testing fixtures. You can merely have an empty file.

File: files/default/lib/packer_pipeline.rb [hunk 11]
```diff
@@ -0,0 +1,80 @@
...
+      # .dig fails if it tries to index an array with a string, which is ridiculous.
+      # This will fail on other json files that start with in an array.
+      # I suppose this helps more in testing than anything, since there aren't going
+      # to be other json files in the repo.
+      next unless t_data.class.name == 'Hash'
+
+      scripts = t_data.dig('provisioners', 0, 'scripts')
+
+      # .dig returns nil if it doesn't find that path in the hash.
+      next if scripts.nil?
```
Comment (ramereth) [inline files/default/lib/packer_pipeline.rb]: You should add a test which verifies this.

File: files/default/lib/packer_pipeline.rb [hunk 12]
```diff
@@ -0,0 +1,80 @@
...
+  # If it receives a script, then it locates all
+  # templates that use that script and returns them in an array.
+  def self.find_dependent_templates(file)
+    return [File.basename(file)] if file =~ /.json$/
+
+    file = File.basename file
+    dependent_templates = []
+
+    # go to dir containing all images
+    Dir.chdir(ENV['PACKER_TEMPLATES_DIR'] || './bento/packer')
```
Comment (ramereth) [inline files/default/lib/packer_pipeline.rb]: This default should be changed as we're no longer using that other repo.

File: files/default/lib/packer_pipeline.rb (hunk 8 above)
Comment (knightsamar) [inline files/default/lib/packer_pipeline.rb]: Explained a bit more and also added a gist link for example explaining what is happening, in the commit comment.

**Comment by knightsamar** [comment]
All reviewed things fixed. Now merging this with my admin powers in interest of time and deadlines.

**Comment by knightsamar** [comment]
!bump minor ~

**Comment by osuosl-manatee** [comment]
Jenkins has merged this PR into `master` and has automatically performed a minor-level version bump to v1.8.0.  Have a nice day!

**Pull request merged**